tools/twix/src/repository_parameters.rs

```diff
@@ -32,46 +32,61 @@ impl RepositoryParameters {
         let repository = self.repository.clone();
-        let Ok(head_id) = self
-            .head_id_from_address(address)
+        let Ok(hardware_ids) = self
+            .hardware_ids_from_address(address)
         else {
             error!("failed to get head ID from address {address}");
             return
         };
         self.runtime.spawn(async move {
-            let mut stored_value = repository
-                .read_configuration(&head_id)
+            let stored_complete_parameter_tree: Value = parameters::directory::deserialize(
+                repository.root_directory(),
+                &hardware_ids.body_id,
+                &hardware_ids.head_id,
+            )
+            .await
+            .unwrap_or_default();
+
+            // value is just the "leaf" of the "path", make the tree structure to diff/ merge against complete parameter tree.
+            let supplied_value_as_sparse_tree =
+                make_sparse_value_tree_from_path(path.as_str(), &value);
+            let diff = get_diff_against_stored_value(
+                &stored_complete_parameter_tree,
+                &supplied_value_as_sparse_tree,
+            );
+            let mut head_parameters = repository
+                .read_configuration(&hardware_ids.head_id)
                 .await
                 .unwrap_or_default();
-
-            let nested_value_to_be_added = nest_value_at_path(path.as_str(), &value);
-
-            merge_json(&mut stored_value, &nested_value_to_be_added);
+            merge_json(&mut head_parameters, &diff);
 
             if let Err(error) = repository
-                .write_configuration(&head_id, &stored_value)
+                .write_configuration(&hardware_ids.head_id, &head_parameters)
                 .await
             {
                 error!("Failed to write value to repository: {error:#?}");
             }
         });
     }
 
-    fn head_id_from_address(&self, address: &str) -> Result<String> {
+    fn hardware_ids_from_address(&self, address: &str) -> Result<HardwareIds> {
         if address == "localhost" {
-            return Ok("webots".to_string());
+            return Ok(HardwareIds {
+                body_id: "webots".to_string(),
+                head_id: "webots".to_string(),
+            });
         }
         let nao_number =
             last_octet_from_ip_address(address.parse().wrap_err("failed to parse IP address")?);
-        self.head_id_from_nao_number(nao_number)
+        self.hardware_ids_from_nao_number(nao_number)
             .wrap_err_with(|| format!("failed to get head ID from NAO number {nao_number}"))
     }
 
-    fn head_id_from_nao_number(&self, nao_number: u8) -> Result<String> {
+    fn hardware_ids_from_nao_number(&self, nao_number: u8) -> Result<HardwareIds> {
         self.ids
             .get(&nao_number)
             .ok_or_else(|| eyre!("no IDs known for NAO number {nao_number}"))
-            .map(|id| id.head_id.clone())
+            .cloned()
     }
```

Comment:
Same here: You could return a borrowed value and remove the `.cloned()`. The remaining code only takes references from your newly cloned `HardwareIds` instance. Therefore it does not matter if you already return a borrow.

Reply:
I'll still have to add a `.cloned()` in `hardware_ids_from_address` as that may return a temporary;
```rust
    fn hardware_ids_from_address(&self, address: &str) -> Result<HardwareIds> {
        if address == "localhost" {
            return Ok(HardwareIds { // temporary
                body_id: "webots".to_string(),
                head_id: "webots".to_string(),
            });
        }
        let nao_number =
            last_octet_from_ip_address(address.parse().wrap_err("failed to parse IP address")?);
        self.hardware_ids_from_nao_number(nao_number)
            .wrap_err_with(|| format!("failed to get head ID from NAO number {nao_number}"))
            .cloned() // clone
    }
```

Reply:
Yes, I overlooked this. Then it needs to stay as is.